src/app/GitUI/Editor/Diff/DiffHighlightService.cs

```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
-    /// <returns>The last line number processed.</returns>
-    protected virtual int TryHighlightAddedAndDeletedLines(IDocument document, int line, LineSegment lineSegment)
-        => line;
-
-    protected void ProcessLineSegment(IDocument document, ref int line,
-        LineSegment lineSegment, string prefixStr, Color color, bool invertMatch = false)
+    private void HighlightAddedAndDeletedLines(IDocument document)
     {
-        if (!DoesLineStartWith(document, lineSegment.Offset, prefixStr, invertMatch))
+        foreach (ISegment segment in GetAllLines(DiffLineType.Minus))
         {
-            return;
+            document.MarkerStrategy.AddMarker(CreateTextMarker(segment.Offset, segment.Length, _removedBackColor));
         }
 
-        LineSegment endLine = document.GetLineSegment(line);
+        foreach (ISegment segment in GetAllLines(DiffLineType.Plus))
+        {
+            document.MarkerStrategy.AddMarker(CreateTextMarker(segment.Offset, segment.Length, _addedBackColor));
+        }
 
-        for (;
-            line < document.TotalNumberOfLines
-            && DoesLineStartWith(document, endLine.Offset, prefixStr, invertMatch);
-            line++)
+        foreach (ISegment segment in GetAllLines(DiffLineType.Header))
         {
-            endLine = document.GetLineSegment(line);
+            document.MarkerStrategy.AddMarker(CreateTextMarker(segment.Offset, segment.Length, AppColor.DiffSection.GetThemeColor()));
         }
+    }
 
-        line = Math.Max(0, line - 2);
-        endLine = document.GetLineSegment(line);
+    /// <summary>
+    ///  Matches related removed and added lines in a consecutive block of a patch document and marks identical parts dimmed.
+    /// </summary>
+    private void MarkInlineDifferences(IDocument document)
+    {
+        int index = 0;
+        List<DiffLineInfo> diffLines = [.. _diffLinesInfo.DiffLines.Values.OrderBy(i => i.LineNumInDiff)];
+        bool found = false;
 
-        document.MarkerStrategy.AddMarker(new TextMarker(lineSegment.Offset,
-            (endLine.Offset + endLine.TotalLength) -
-            lineSegment.Offset, TextMarkerType.SolidBlock, color,
-            ColorHelper.GetForeColorForBackColor(color)));
+        const int diffContentOffset = 1; // in order to skip the prefixes '-' / '+'
+        MarkerStrategy markerStrategy = document.MarkerStrategy;
 
-        return;
+        // Process the next blocks of removed / added diffLines and mark in-line differences
+        while (index < diffLines.Count)
+        {
+            found = false;
 
-        bool DoesLineStartWith(IDocument document, int offset, string prefixStr, bool invertMatch)
-            => invertMatch ^ LinePrefixHelper.DoesLineStartWith(document, offset, prefixStr);
-    }
+            // git-diff presents the removed lines directly followed by the added in a "block"
+            IReadOnlyList<ISegment> linesRemoved = GetBlockOfLines(diffLines, DiffLineType.Minus, ref index, ref found);
```

Comment:
`found` is not evaluated.

Reply:
It passes information from minus to plus, if a new block is started.
Inline markers are only considered for minus directly followed by plus lines
This could be rewritten to remove found and continue if there are no linesRemoved

Reply:
I changed this. Note that the code increases in size (a few lines) and is not more efficient. But slightly easier to understand.